src/picongpu/include/plugins/particleCalorimeter/ParticleCalorimeter.hpp

```diff
@@ -0,0 +1,530 @@
+            guardSuperCells * SuperCellSize::toRT());
+
+        /* kernel call */
+        algorithm::kernel::Foreach<SuperCellSize> foreach;
+        foreach(zone, cursor::make_MultiIndexCursor<simDim>(), particleCalorimeterKernel);
+
+        /* copy to host */
+        *this->hBufCalorimeter = *this->dBufCalorimeter;
+
+        /* mpi reduce */
+        using namespace lambda;
+        (*this->allGPU_reduce)(*this->hBufTotalCalorimeter, *this->hBufCalorimeter, _1 + _2);
+        if(!this->allGPU_reduce->root())
+            return;
+
+        this->writeToHDF5File(currentStep);
+    }
+
+
+    void setMappingDescription(MappingDesc* cellDescription)
+    {
+        this->cellDescription = cellDescription;
+    }
+
+
+    void pluginRegisterHelp(po::options_description& desc)
+    {
+        desc.add_options()
+        ((this->prefix + ".period").c_str(), po::value<uint32_t > (&this->notifyPeriod)->default_value(0),
+            "enable plugin [for each n-th step]")
+        ((this->prefix + ".numBinsYaw").c_str(), po::value<uint32_t > (&this->numBinsYaw)->default_value(64),
+            "number of bins for angle yaw.")
+        ((this->prefix + ".numBinsPitch").c_str(), po::value<uint32_t > (&this->numBinsPitch)->default_value(64),
+            "number of bins for angle pitch.")
+        ((this->prefix + ".numBinsEnergy").c_str(), po::value<uint32_t > (&this->numBinsEnergy)->default_value(1),
+            "number of bins for the energy spectrum. Disabled by default.")
+        ((this->prefix + ".minEnergy").c_str(), po::value<float_X > (&this->minEnergy)->default_value(float_X(0.0)),
+            "minimal detectable energy in keV.")
+        ((this->prefix + ".maxEnergy").c_str(), po::value<float_X > (&this->maxEnergy)->default_value(float_X(1.0e3)),
+            "maximal detectable energy in keV.")
+        ((this->prefix + ".logScale").c_str(), po::bool_switch(&this->logScale),
+            "enable logarithmic energy scale.")
+        ((this->prefix + ".openingYaw").c_str(), po::value<float_X > (&this->openingYaw_deg)->default_value(float_X(360.0)),
+            "opening angle yaw in degrees. 0 < x < 360.")
+        ((this->prefix + ".openingPitch").c_str(), po::value<float_X > (&this->openingPitch_deg)->default_value(float_X(180.0)),
+            "opening angle pitch in degrees. 0 < x < 180.")
+        ((this->prefix + ".posYaw").c_str(), po::value<float_64 > (&this->posYaw_deg)->default_value(float_64(0.0)),
+            "yaw coordinate of calorimeter position in degrees. Defaults to +y direction.")
+        ((this->prefix + ".posPitch").c_str(), po::value<float_64 > (&this->posPitch_deg)->default_value(float_64(0.0)),
+            "pitch coordinate of calorimeter position in degrees. Defaults to +y direction.");
+    }
+
+
+    std::string pluginGetName() const
+    {
+        return this->name;
+    }
+
+
+    void onParticleLeave(const std::string& speciesName, int32_t direction)
+    {
+        if(this->notifyPeriod == 0)
```

Comment:
that is an interesting question. should we ignore steps or should we store those particles beforehand? ;)
maybe someone wants to restart and enable the plugin later on :D (can be solved later)